contracts/PetPark.sol

```diff
@@ -1,7 +1,133 @@
+        /** Only owners of contract can set animal park */
+        require(msg.sender == owner, "Not an owner");
+        animalPark[animal] = count;
+        emit Added(animal, count);
+    }
+
+    function animalCounts(AnimalType animal) public view returns (uint256) {
+        return animalPark[animal];
+    }
+
+    /**
+    Takes Age, Gender and Animal Type.
+    Can borrow only one animal at a time. Use function giveBackAnimal to borrow another animal.
+    Men can borrow only Dog and Fish.
+    Women can borrow every kind, but women aged under 40 are not allowed to borrow a Cat.
+    Throw an error if an address has called this function before using other values for Gender and Age.
+    Emit event Borrowed with parameter Animal Type.
+     */
+
+    function borrow(
+        uint256 age,
+        Gender gender,
+        AnimalType animal
+    ) public validAnimal(animal) {
+        /** Age must be greater then zero */
+        require(age > 0, "Invalid Age");
+
+        /** Animal must be in park */
+        require(animalPark[animal] > 0, "Selected animal not available");
+
+        Borrower memory myBorrow = borrowers[msg.sender];
+
+        /** If borrower has already adopted a pet */
+        if (myBorrow.age != 0) {
+            /** Check if function has been called before with other values for 
+             Gender and Age */
+            require(age == myBorrow.age, "Invalid Age");
+            require(gender == myBorrow.gender, "Invalid Gender");
+        }
+
+        /**  Check if user has already adopted a pet and throw error */
+        require(myBorrow.animal == AnimalType.None, "Already adopted a pet.");
+
+        if (gender == Gender.Male) {
+            /** If gender is male, only allow Dogs, and Fish  */
+            require(
+                gender == Gender.Male &&
+                    (animal == AnimalType.Dog || animal == AnimalType.Fish),
+                "Invalid animal for men"
+            );
+        }
+
+        if (gender == Gender.Female && age < 40) {
+            /** if Female and under 40 don't allow cats */
+            require(
+                animal != AnimalType.Cat,
+                "Invalid animal for women under 40"
+            );
+        }
+
+        Borrower memory borrower = Borrower(animal, age, gender);
+        borrowers[msg.sender] = borrower;
+        animalPark[animal] = animalPark[animal] - 1;
+        emit Borrowed(animal);
+    }
 
-}
\ No newline at end of file
+    /**
+    Throw an error if user hasn't borrowed before.
+    Emit event Returned with parameter Animal Type.
+     */
+    function giveBackAnimal() public {
+        Borrower memory myBorrow = borrowers[msg.sender];
+        /** If they haven't borrowed a pet throw error */
+        require(myBorrow.age != 0, "No borrowed pets");
+        animalPark[myBorrow.animal] = animalPark[myBorrow.animal] + 1;
+        borrowers[msg.sender] = Borrower(AnimalType.None, 0, Gender.Male);
```

Comment:
Setting this borrower back to empty values. So they can borrow again.

Reply:
You can use `delete borrowers[msg.sender]` instead. It's better because it releases storage unlike the code above which still uses the storage.

Reply:
Ok I'll change that!